src/dinit-iostream.cc

```diff
@@ -0,0 +1,720 @@
+        }
+        else if (r < 0) {
+            io_error = errno;
+            return -1;
+        }
+    }
+    char result = *buf->get_ptr(0);
+    buf->consume(1);
+    return result;
+}
+
+int istream::getc()
+{
+    int r = getc_nx();
+    if (r < 0) {
+        throw_exception_on(io_states::eofbit | io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+ssize_t istream::get_line_nx(std::string &dest, char delim) noexcept
+{
+    if (!good()) {
+        return -1;
+    }
+    if (buf->get_length() == 0) {
+        int r = load_into_buf(buf->get_size());
+        if (r == 0) {
+            eof_state = true;
+            return 0;
+        }
+        if (r < 0) {
+            io_error = errno;
+            return -1;
+        }
+    }
+    dest.clear();
+
+    char *ptr = buf->get_ptr(0);
+    unsigned len = buf->get_contiguous_length(ptr);
+    char *eptr = static_cast<char *>(std::memchr(ptr, delim, len));
+
+    if (eptr == ptr) {
+        buf->consume(1);
+        return 1; // An empty line
+    }
+    while (eptr == nullptr) {
+        // If it hasn't any delim; We try to read as much as possible to reach the EOF or find a
+        // "delim" character.
+        try {
+            // dest.append() may throw out_of_range, length_error or bad_alloc
+            dest.append(ptr, len);
+        }
+        catch (const std::exception &e) {
+            string_failed = true;
+            return -1;
+        }
+        buf->consume(len);
+
+        if (buf->get_length() == 0) {
+            int r = load_into_buf(buf->get_size());
+            if (r == 0) {
+                return dest.size();
+            }
+            if (r < 0) {
+                io_error = errno;
+                return -1;
+            }
+        }
+
+        ptr = buf->get_ptr(0);
+        len = buf->get_contiguous_length(ptr);
+        eptr = static_cast<char *>(std::memchr(ptr, delim, len));
+    }
+    unsigned index = eptr - ptr;
+    try {
+        // dest.append() may throw out_of_range, length_error or bad_alloc
+        dest.append(ptr, index);
+    }
+    catch (const std::exception &e) {
+        string_failed = true;
+        return -1;
+    }
+    buf->consume(index + 1); // Consume the delimiter character as well
+    return dest.size();
+}
+
+ssize_t istream::get_line(std::string &dest, char delim)
+{
+    ssize_t r = get_line_nx(dest, delim);
+    if (r <= 0) {
+        throw_exception_on(io_states::eofbit | io_states::buffer_failbit | io_states::string_failbit
+                | io_states::io_failbit);
+    }
+    return r;
+}
+
+ssize_t istream::get_line_until_eof(std::string &dest, char delim)
+{
+    ssize_t r = get_line_nx(dest, delim);
+    if (r < 0) {
+        throw_exception_on(io_states::buffer_failbit | io_states::string_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+istream::operator bool() noexcept
+{
+    return good();
+}
+
+istream::~istream() noexcept
+{
+    if (!is_open()) return;
```

Comment:
This check couldn't be deleted because passing invalid fd to `bp_sys::close()` (in `test-bpsys.cc`) will abort the whole thing.

Will we delete this check and fix `test-bpsys.cc` or keep it?

Reply:
Hmm, let's just keep this check. Since it's in the destructor it's more reasonable anyway and setting the errno value isn't useful.